File: kernel/internal/mcp/mcp_router.py
import time
from typing import Dict, Any, Callable, List, Optional

class MCPToolDefinition:
    def __init__(self, name: str, description: str, server_id: str, parameters: Dict[str, Any], handler: Callable):
        self.name = name
        self.description = description
        self.server_id = server_id
        self.parameters = parameters
        self.handler = handler

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "description": self.description,
            "server_id": self.server_id,
            "parameters": self.parameters
        }


class MCPRouter:
    """Model Context Protocol (MCP) Transport Router & Dispatcher"""
    def __init__(self):
        self.tools: Dict[str, MCPToolDefinition] = {}
        self.servers: Dict[str, Dict[str, Any]] = {}

    def register_server(self, server_id: str, name: str, transport_type: str = "in-memory"):
        self.servers[server_id] = {
            "server_id": server_id,
            "name": name,
            "transport": transport_type,
            "status": "connected",
            "registered_at": time.time(),
            "tools_count": 0
        }

    def register_tool(self, tool_def: MCPToolDefinition):
        self.tools[tool_def.name] = tool_def
        if tool_def.server_id in self.servers:
            self.servers[tool_def.server_id]["tools_count"] += 1

    def list_tools(self) -> List[Dict[str, Any]]:
        return [t.to_dict() for t in self.tools.values()]

    def list_servers(self) -> List[Dict[str, Any]]:
        return list(self.servers.values())

    def dispatch(self, tool_name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
        start = time.time()
        if tool_name not in self.tools:
            return {
                "error": f"Tool '{tool_name}' not found in MCP registry",
                "result": None,
                "latency_ms": round((time.time() - start) * 1000, 2)
            }

        tool = self.tools[tool_name]
        try:
            res = tool.handler(arguments)
            latency = round((time.time() - start) * 1000, 2)
            return {
                "error": None,
                "tool_name": tool_name,
                "server_id": tool.server_id,
                "result": res,
                "latency_ms": latency
            }
        except Exception as e:
            latency = round((time.time() - start) * 1000, 2)
            return {
                "error": str(e),
                "tool_name": tool_name,
                "server_id": tool.server_id,
                "result": None,
                "latency_ms": latency
            }

File: kernel/internal/mcp/mcp_router.go
package mcp

import (
	"fmt"
	"sync"
	"time"
)

type ServerInfo struct {
	ServerID     string `json:"server_id"`
	Name         string `json:"name"`
	Transport    string `json:"transport"`
	Status       string `json:"status"`
	RegisteredAt int64  `json:"registered_at"`
	ToolsCount   int    `json:"tools_count"`
}

type ToolHandler func(args map[string]interface{}) (interface{}, error)

type ToolInfo struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	ServerID    string                 `json:"server_id"`
	Parameters  map[string]interface{} `json:"parameters"`
	Handler     ToolHandler            `json:"-"`
}

type Router struct {
	mu      sync.RWMutex
	servers map[string]*ServerInfo
	tools   map[string]ToolInfo
}

func NewRouter() *Router {
	return &Router{
		servers: make(map[string]*ServerInfo),
		tools:   make(map[string]ToolInfo),
	}
}

func (r *Router) RegisterServer(serverID, name, transport string) {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.servers[serverID] = &ServerInfo{
		ServerID:     serverID,
		Name:         name,
		Transport:    transport,
		Status:       "connected",
		RegisteredAt: time.Now().Unix(),
		ToolsCount:   0,
	}
}

func (r *Router) RegisterTool(info ToolInfo) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.tools[info.Name] = info
	if server, ok := r.servers[info.ServerID]; ok {
		server.ToolsCount++
	}
	return nil
}

func (r *Router) ListTools() []ToolInfo {
	r.mu.RLock()
	defer r.mu.RUnlock()

	res := make([]ToolInfo, 0, len(r.tools))
	for _, t := range r.tools {
		res = append(res, t)
	}
	return res
}

func (r *Router) ListServers() []ServerInfo {
	r.mu.RLock()
	defer r.mu.RUnlock()

	res := make([]ServerInfo, 0, len(r.servers))
	for _, s := range r.servers {
		res = append(res, *s)
	}
	return res
}

func (r *Router) Dispatch(toolName string, args map[string]interface{}) (interface{}, error) {
	r.mu.RLock()
	tool, exists := r.tools[toolName]
	r.mu.RUnlock()

	if !exists {
		return nil, fmt.Errorf("tool %s not registered", toolName)
	}

	if tool.Handler != nil {
		return tool.Handler(args)
	}

	return fmt.Sprintf("Executed tool %s on server %s with args %v", tool.Name, tool.ServerID, args), nil
}
